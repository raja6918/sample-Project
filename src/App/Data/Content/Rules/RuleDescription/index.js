import RuleDescriptionWithTranslation from './RuleDescription';

export default RuleDescriptionWithTranslation;
