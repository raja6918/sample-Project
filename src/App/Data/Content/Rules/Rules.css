.MuiDrawer-paper {
  overflow-y: hidden;
}
.newRulecls .MuiFormLabel-root {
  font-size: 0.8rem;
}
.newRulecls label + .MuiInput-formControl {
  margin-top: 0;
}
.newRulecls .MuiInputLabel-formControl {
  transform: none;
}
.drawerCls .MuiPaper-root {
  background-color: #efefef;
}
.more-option .MuiPaper-elevation8 {
  box-shadow: 0px 5px 5px -3px rgba(0, 0, 0, 0.2) !important;
}
.titleCls {
  top: 6px;
  position: absolute;
  left: -8px;
  font-size: 1.1rem;
  max-width: 235px;
}
.subTitleCls {
  position: absolute;
  left: -8px;
  font-size: 1.2em;
  font-weight: 400;
  line-height: 1.2;
}
.rootSubTitleCls {
  position: absolute;
  left: 39px;
  font-size: 1.2em;
  font-weight: 400;
}
.basicTtlcls {
  font-size: 1.1rem;
  position: absolute;
  left: 47px;
  top: 12px;
  max-width: 235px;
}
.basicImg {
  left: 14px;
  width: 21px;
  position: absolute;
  top: 24px;
}
.displayImg {
  display: none;
}

.ruleSetClass {
  width: 219px;
  left: 10px;
}
.ruleSetClass label {
  font-size: 0.9rem;
}
.container-cls {
  padding-bottom: 5px;
}
.ruleset-select {
  min-width: 217px;
}
.tree-icon {
  z-index: 999;
}
.ruleset-tooltip {
  overflow: hidden;
  text-overflow: ellipsis;
  display: block !important;
}
