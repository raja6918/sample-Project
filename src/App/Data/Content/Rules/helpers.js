export const checkEmptyFilter = filters => {
  return typeof filters === 'object'
    ? Object.values(filters).filter(a => {
        const val = /^ *$/.test(a);
        return !val;
      }).length
    : 0;
};
