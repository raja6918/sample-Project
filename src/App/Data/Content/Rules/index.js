import RulesWithErrorHandling from './Rules';

export default RulesWithErrorHandling;
