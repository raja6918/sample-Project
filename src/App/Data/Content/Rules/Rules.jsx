import React, { Component, createRef, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Grid, MenuItem, Select } from '@material-ui/core';
import DataNotFound from '../../../../components/DataNotFound';
import Container from '../../../../components/DataContent/Container';
import GenericCollapsibleTable from '../../../../components/GenericCollapsibleTable';
import AddHeader from '../../../../components/Headers/AddHeader';
import RedirectLink from '../../../../components/RedirectLink';
import EditModeBar from '../../../../components/EditModeBar';
import { orderBy, order, getHeaders, INITIAL_ITEMS } from './constants';
import storage from '../../../../utils/storage';
import RuleDescription from './RuleDescription';
import ModalLoader from '../../../../components/ModalLoader';
import * as ruleService from '../../../../services/Data/rules';
import MenuText from '../../../../components/Menu/MenuText';
import withErrorHandler from '../../../../components/ErrorHandler/withErrorHandler';
import { connect } from 'react-redux';
import { checkEmptyFilter } from './helpers.js';
import { prepareSortPayload } from './../../../../components/GenericCollapsibleTable/helpers';
import AccessEnabler from '../../../../components/AccessEnabler';
import scopes from '../../../../constants/scopes';
import { checkPermission } from '../../../../utils/common';
import { pushRuleChangeDataToAnalytics } from '../../../../utils/analytics';
import { clearMasterData } from '../../../../actions/data';

export class Rules extends Component {
  state = {
    rules: [],
    ruleSets: [],
    totalDataSize: 0,
    ruleDescriptions: [],
    currentRuleSetId: storage.getItem('rulesetId'),
    userId: this.props.userData.id,
    filteredDataSize: 0,
    openLoader: false,
    titleLoader: this.props.t('DATA.rules.loader'),
    isFilter: false,
    isLoading: false,
    filters: null,
  };

  rulesPayload = {
    startIndex: 0,
    endIndex: INITIAL_ITEMS,
    sort: prepareSortPayload({ field: orderBy }, order),
  };

  tableRef = createRef();

  async componentDidMount() {
    try {
      const { openItemId } = this.props;
      const { currentRuleSetId } = this.state;
      let tempRulesetId = currentRuleSetId;
      this.setState({ isLoading: true });
      await this.getAllRuleSets();
      if (currentRuleSetId) {
        this.getAllRules();
      } else {
        ruleService
          .getDefaultCrewgroup(openItemId)
          .then(result => {
            if (Array.isArray(result.data) && result.data[0].ruleset) {
              tempRulesetId = result.data[0].ruleset;
            } else {
              const parentRuleSet = this.state.ruleSets.find(
                item => item.fallback === 0
              );
              tempRulesetId = parentRuleSet ? parentRuleSet.id : null;
            }
            if (tempRulesetId)
              this.setState(
                { currentRuleSetId: tempRulesetId },
                this.getAllRules
              );
          })
          .catch(error => {
            if (error.response) this.props.reportError({ error });
          });
      }
    } catch (e) {
      console.error(e);
      this.setState({ isLoading: false });
    }
  }

  componentWillUnmount() {
    this.props.clearMasterData();
  }

  getAllRuleSets = () => {
    const { userId } = this.state;
    const { openItemId } = this.props;
    return ruleService
      .getRuleSets(openItemId, userId)
      .then(ruleSets => {
        this.setState({ ruleSets });
      })
      .catch(error => {
        if (error.response) this.props.reportError({ error });
      });
  };

  getAllRules = () => {
    const { openItemId } = this.props;
    const { currentRuleSetId, userId } = this.state;
    ruleService
      .getRules(userId, openItemId, currentRuleSetId, this.rulesPayload)
      .then(response => {
        this.setState({
          rules: response.data,
          totalDataSize: response.totalDataSize,
        });
      })
      .catch(error => {
        if (error.response) this.props.reportError({ error });
      })
      .finally(() => {
        this.setState({ isLoading: false });
      });
  };

  getCurrentRuleSetName = () => {
    const { ruleSets, currentRuleSetId } = this.state;

    const tempRuleSet = ruleSets.find(item => item.id === currentRuleSetId);
    return tempRuleSet ? tempRuleSet.name : '';
  };

  onChangeRuleSet = event => {
    const { openItemId } = this.props;
    const { userId } = this.state;
    this.rulesPayload = {
      ...this.rulesPayload,
      startIndex: 0,
      endIndex: INITIAL_ITEMS,
    };
    ruleService
      .getRules(userId, openItemId, event.target.value, this.rulesPayload)
      .then(response => {
        this.setState({
          rules: response.data,
          currentRuleSetId: event.target.value,
          ruleDescriptions: [],
          filteredDataSize: response.totalDataSize,
        });
        // Reset table body ie to collapse already opened rows
        if (this.tableRef.current) {
          this.tableRef.current.resetRows();
          this.tableRef.current.moveScrollBarToTop();
        }
        storage.setItem('rulesetId', event.target.value);
      })
      .catch(error => {
        if (error.response) this.props.reportError({ error });
      });
  };

  handleStateChange = async (value, data) => {
    const rules = this.state.rules;
    const ruleIndex = rules.findIndex(rule => rule.code === data.code);
    try {
      // Call the change state API
      await ruleService.activateRule(
        this.state.userId,
        this.props.openItemId,
        data.ruleset,
        data.code,
        value
      );
      if (ruleIndex !== -1) {
        rules[ruleIndex].active = value;
        rules[ruleIndex].activeDefinedInCurrent = true;
      }
    } catch (error) {
      if (error.response) {
        this.props.reportError({ error });
      }

      // In case of error reset the selectbox
      if (ruleIndex !== -1) {
        rules[ruleIndex].active = !value;
      }
    } finally {
      this.setState({ rules }, () => {
        this.filterStates();
      });
    }
  };

  handleStateReset = async data => {
    try {
      // Call the revert rule API
      await ruleService.revertRule(
        this.state.userId,
        this.props.openItemId,
        data.ruleset,
        data.code
      );

      // Once revert is success, Call rule detail API to fetch it's previous value
      const ruleDetails = await this.fetchRuleDetails(data.code, data.ruleset);

      const rules = this.state.rules;
      const ruleIndex = rules.findIndex(rule => rule.code === data.code);
      if (ruleIndex !== -1) {
        rules[ruleIndex].active = ruleDetails.active;
        rules[ruleIndex].activeDefinedInCurrent =
          ruleDetails.activeDefinedInCurrent;
      }

      this.setState({ rules }, () => {
        this.filterStates();
      });
    } catch (error) {
      if (error.response) this.props.reportError({ error });
    }
  };

  fetchRuleDetails = async (code, ruleSet) => {
    try {
      return await ruleService.getRuleDescription(
        code,
        this.props.openItemId,
        this.state.userId,
        ruleSet
      );
    } catch (error) {
      if (error.response) this.props.reportError({ error });
    }
  };

  /**
   * Method to remove rule in rules table if user apply state filter and then change state
   */
  filterStates = () => {
    const { filters, isFilter } = this.state;
    if (isFilter) {
      const active = filters ? filters.active : null;
      if (typeof active === 'boolean') {
        const rules = this.state.rules.filter(rule => rule.active === active);
        this.setState({ rules });
      }
    }
  };

  handleDisable = (value, data) => {
    const { readOnly, userPermissions } = this.props;
    return (
      data.ruleTypeName === 'Definition' ||
      readOnly ||
      !checkPermission(scopes.rules.ruleEdit, userPermissions)
    );
  };

  highlighter = (value, data) => {
    if (value) {
      return value === this.getCurrentRuleSetName();
    }
    return false;
  };

  checkStateToolTipStatus = data =>
    !data.activeDefinedInCurrent || !data.ruleHasReject;

  handleStateTooltipDisable = (value, data) => {
    const { readOnly, userPermissions } = this.props;
    return (
      this.checkStateToolTipStatus(data) ||
      readOnly ||
      !checkPermission(scopes.rules.ruleEdit, userPermissions)
    );
  };

  getTooltipContent = (value, data) =>
    this.checkStateToolTipStatus(data)
      ? this.props.t('DATA.rules.tooltip.settingsFrom', {
          name: data.activeReferenceRulesetName,
        })
      : this.props.t('DATA.rules.tooltip.revertTo', {
          name: data.activeReferenceRulesetName,
        });

  /**
   * This method will be called once the API call is successfully completed by Description component
   */
  handleParamSet = async (value, data, code) => {
    this.toggleLoader(true);
    const ruleDescriptions = this.state.ruleDescriptions.map(description => {
      if (description.code === code && data.name) {
        const paramObject = description.userDescription.references[data.name];
        if (paramObject) {
          paramObject.definedInCurrent = true;
          paramObject.value = value;
          description.userDescription.references[data.name] = paramObject;
          const payload = {};
          payload['iRuleCategory'] = description.ruleCategoryName;
          payload['iRuleType'] = description.ruleTypeName;
          pushRuleChangeDataToAnalytics(payload);
        }
      }
      return description;
    });

    // If paramater change we need to change the Setting From value too.
    const selectedRuleSet = this.state.ruleSets.find(
      item => item.id === this.state.currentRuleSetId
    );

    const rules = this.state.rules;
    const ruleIndex = rules.findIndex(rule => rule.code === code);
    if (ruleIndex !== -1) {
      rules[ruleIndex].settingsFromRulesetName = selectedRuleSet
        ? selectedRuleSet.name
        : rules[ruleIndex].settingsFromRulesetName;
    }

    this.setState({ ruleDescriptions, rules }, () => {
      this.refreshRules();
    });
  };

  handleParamReset = async (data, ruleset, code) => {
    try {
      this.toggleLoader(true);
      // Call the revert param API
      await ruleService.revertParam(
        this.state.userId,
        this.props.openItemId,
        ruleset,
        data.name
      );

      this.refreshRules();
    } catch (error) {
      if (error.response) this.props.reportError({ error });
    }
  };

  userDescriptionTransformer = data => {
    const transformedData = {};
    const references = {};
    transformedData.data = data.data;
    data.references.forEach(reference => {
      if (reference.name) references[reference.name] = reference;
    });
    transformedData.references = references;
    return transformedData;
  };

  fetchDescription = async (data, ruleSet) => {
    try {
      const checkExist = this.state.ruleDescriptions.find(
        description => description.code === data.code
      );

      // If it is not in global ruleDescriptions state, Then fetch it via API
      if (!checkExist) {
        // calling API
        const description = await ruleService.getRuleDescription(
          data.code,
          this.props.openItemId,
          this.state.userId,
          data.ruleset || ruleSet
        );
        description.userDescription = this.userDescriptionTransformer(
          description.userDescription
        );
        if (description) {
          this.setState(prevState => ({
            ruleDescriptions: prevState.ruleDescriptions.concat(description),
          }));
        }
      }
    } catch (error) {
      if (error.response) this.props.reportError({ error });
    }
  };

  handleSort = sortPayload => {
    const { openItemId } = this.props;
    const { currentRuleSetId, userId } = this.state;
    this.rulesPayload = {
      ...this.rulesPayload,
      startIndex: 0,
      endIndex: INITIAL_ITEMS,
      sort: sortPayload,
    };

    ruleService
      .getRules(userId, openItemId, currentRuleSetId, this.rulesPayload)
      .then(response => {
        this.setState({ rules: response.data });
        // Reset table body ie to collapse already opened rows
        if (this.tableRef.current) {
          this.tableRef.current.resetRows();
          this.tableRef.current.moveScrollBarToTop();
        }
      });
  };

  handleFilter = filters => {
    const isEmpTyFilterObj = checkEmptyFilter(filters);
    const { openItemId } = this.props;
    const { currentRuleSetId, userId } = this.state;
    this.rulesPayload = {
      ...this.rulesPayload,
      filterCriteria: filters,
      startIndex: 0,
      endIndex: INITIAL_ITEMS,
    };
    ruleService
      .getRules(userId, openItemId, currentRuleSetId, this.rulesPayload)
      .then(response => {
        this.setState({
          rules: response.data,
          isFilter: isEmpTyFilterObj ? true : false,
          filteredDataSize: response.totalDataSize,
          filters: isEmpTyFilterObj ? filters : null,
        });
      })
      .catch(err => {
        this.setState({
          rules: [],
          isFilter: isEmpTyFilterObj ? true : false,
          filteredDataSize: 0,
          filters: isEmpTyFilterObj ? filters : null,
        });
        console.error(err);
        if (err.response) this.props.reportError({ err });
      });
  };

  fetchData = nextItems => {
    const {
      currentRuleSetId,
      userId,
      totalDataSize,
      isFilter,
      filteredDataSize,
    } = this.state;
    const { openItemId } = this.props;
    const { endIndex: currentEndIndex } = this.rulesPayload;

    if (
      currentEndIndex === totalDataSize ||
      (totalDataSize > 0 && currentEndIndex > totalDataSize) ||
      (isFilter && currentEndIndex > filteredDataSize)
    )
      return;

    const nextStartIndex = currentEndIndex;
    let nextEndIndex = currentEndIndex + nextItems;

    if (nextEndIndex > totalDataSize) {
      nextEndIndex = totalDataSize;
    }

    this.rulesPayload = {
      ...this.rulesPayload,
      startIndex: nextStartIndex,
      endIndex: nextEndIndex,
    };

    ruleService
      .getRules(userId, openItemId, currentRuleSetId, this.rulesPayload)
      .then(response => {
        const newRulesData = [...this.state.rules, ...response.data];
        this.setState({ rules: newRulesData });
      })
      .catch(error => {
        if (error.response) this.props.reportError({ error });
      });
  };

  toggleLoader = open => {
    this.setState({ openLoader: open });
  };

  getRuleCodes = () => {
    return this.state.ruleDescriptions.map(desc => desc.code);
  };

  refreshRules = async () => {
    try {
      this.toggleLoader(true);
      const { openItemId } = this.props;
      const { currentRuleSetId, userId, rules } = this.state;
      const getRules = ruleService.getRules(
        userId,
        openItemId,
        currentRuleSetId,
        { ...this.rulesPayload, startIndex: 0, endIndex: rules.length }
      );
      const getRuleDescriptionFromCodes = ruleService.getRuleDescriptionFromCodes(
        this.getRuleCodes(),
        openItemId,
        userId,
        currentRuleSetId
      );

      const [ruleData, ruleDescriptions] = await Promise.all([
        getRules,
        getRuleDescriptionFromCodes,
      ]);

      ruleDescriptions.forEach(description => {
        description.userDescription = this.userDescriptionTransformer(
          description.userDescription
        );
      });

      this.setState({
        rules: ruleData.data,
        ruleDescriptions: ruleDescriptions,
        totalDataSize: ruleData.totalDataSize,
      });
    } catch (error) {
      if (error.response) this.props.reportError({ error });
    } finally {
      this.toggleLoader(false);
    }
  };

  disableRowStyle = (data, header) => {
    return !data.active && header.field !== 'active' ? { opacity: '0.5' } : {};
  };

  ruleText = 'DATA.rules';

  render() {
    const {
      t,
      editMode,
      openItemId,
      openItemName,
      readOnly,
      ...rest
    } = this.props;
    const {
      rules,
      userId,
      totalDataSize,
      ruleDescriptions,
      ruleSets,
      currentRuleSetId,
      isFilter,
      filteredDataSize,
      isLoading,
    } = this.state;
    return (
      <Fragment>
        <AddHeader
          t={t}
          name={this.getCurrentRuleSetName()}
          editMode={editMode}
          openItemId={openItemId}
          openItemName={openItemName}
          needAddButton={false}
          redirectLink={
            <RedirectLink
              to={{
                pathname: `rules/rule-sets`,
                state: {
                  openItemId: openItemId,
                  openItemName: openItemName,
                  editMode: editMode,
                  readOnly,
                },
              }}
            >
              {t('DATA.ruleSets.link')}
            </RedirectLink>
          }
          readOnly={readOnly}
          {...rest}
        />
        <ModalLoader
          open={this.state.openLoader}
          title={this.state.titleLoader}
          color="white"
        />
        <Container>
          <Grid
            container
            spacing={1}
            alignItems="center"
            className="container-cls"
          >
            <Grid item>{t(`DATA.rules.inputSelector`)} :</Grid>
            <Grid item>
              <Select
                className="ruleset-select"
                onChange={this.onChangeRuleSet}
                value={currentRuleSetId}
                inputProps={{
                  name: name,
                  id: name,
                  required: false,
                }}
              >
                {ruleSets.map(data => (
                  <MenuItem key={data.id} value={data.id}>
                    <MenuText>{data.name ? data.name : ''}</MenuText>
                  </MenuItem>
                ))}
              </Select>
            </Grid>
          </Grid>
          {editMode && <EditModeBar title={t('TEMPLATES.editMode')} />}
          {rules.length || isFilter ? (
            <AccessEnabler
              scopes={scopes.rules.ruleEdit}
              disableComponent
              render={props => (
                <GenericCollapsibleTable
                  headers={getHeaders(
                    t,
                    this.handleStateChange,
                    this.handleDisable,
                    this.highlighter,
                    this.handleStateReset,
                    this.handleStateTooltipDisable,
                    this.getTooltipContent
                  )}
                  data={rules}
                  totalDataSize={totalDataSize}
                  orderBy={orderBy}
                  order={order}
                  handleSort={this.handleSort}
                  handleFilter={this.handleFilter}
                  frontEndFilterdisabled={true}
                  handleFetchData={this.fetchData}
                  filteredDataSize={filteredDataSize}
                  name={t(`${this.ruleText}.name`)}
                  isFilter={isFilter}
                  t={t}
                  editMode={editMode}
                  collapsibleComponent={RuleDescription}
                  collapsibleComponentProps={{
                    userId: userId,
                    readOnly: props.disableComponent || readOnly,
                    scenarioId: openItemId,
                    ruleDescriptions: ruleDescriptions,
                    handleParamSet: this.handleParamSet,
                    handleParamReset: this.handleParamReset,
                    fetchDescription: this.fetchDescription,
                    toggleLoader: this.toggleLoader,
                    refreshRules: this.refreshRules,
                  }}
                  ref={this.tableRef}
                  keyField="code"
                  disableRowStyle={this.disableRowStyle}
                />
              )}
            />
          ) : (
            <Fragment>
              {isLoading ? (
                <DataNotFound
                  text={t('GLOBAL.dataLoading.message', {
                    data: t(`${this.ruleText}.name`).toLowerCase(),
                  })}
                />
              ) : (
                <DataNotFound
                  text={t('GLOBAL.dataNotFound.message', {
                    data: t(`${this.ruleText}.name`).toLowerCase(),
                  })}
                />
              )}
            </Fragment>
          )}
        </Container>
      </Fragment>
    );
  }
}

const mapStateToProps = state => {
  const { user } = state;
  return { userData: user.userData, userPermissions: user.permissions };
};

const mapDispatchToProps = (dispatch, props) => ({
  clearMasterData: () => dispatch(clearMasterData()),
});

const RulesComponent = connect(mapStateToProps, mapDispatchToProps)(Rules);

Rules.propTypes = {
  t: PropTypes.func.isRequired,
  editMode: PropTypes.bool.isRequired,
  openItemId: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  openItemName: PropTypes.string.isRequired,
  reportError: PropTypes.func.isRequired,
  readOnly: PropTypes.bool.isRequired,
  userData: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    email: PropTypes.string,
    role: PropTypes.string,
    userName: PropTypes.string,
  }),
  userPermissions: PropTypes.shape([]).isRequired,
  clearMasterData: PropTypes.func.isRequired,
};

Rules.defaultProps = {
  userData: {},
};

export default withErrorHandler(RulesComponent);
