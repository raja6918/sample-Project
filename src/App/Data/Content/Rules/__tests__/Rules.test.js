import React from 'react';
import { shallow } from 'enzyme';

import { Rules } from '../Rules';

import * as ruleService from '../../../../../services/Data/rules';
import { ruleDescriptionData, rulesData, ruleSets } from '../mockData';

const props = {
  t: msg => msg,
  editMode: false,
  openItemId: '1',
  openItemName: 'Scenario test',
  reportError: jest.fn(),
  userData: { id: 1 },
};

const data = {
  code: ruleDescriptionData[0].code,
  ruleset: ruleDescriptionData[0].ruleset,
};

let wrapper;
beforeAll(() => {
  wrapper = shallow(<Rules {...props} />);
});

test('Rules component render correct', () => {
  expect(wrapper).toMatchSnapshot();
});

test('check whether reportError is called when an error occured in getRuleDescription', async () => {
  ruleService.getRuleDescription = jest.fn(() =>
    // eslint-disable-next-line prefer-promise-reject-errors
    Promise.reject({
      response: {},
    })
  );

  await wrapper.instance().fetchDescription(data);

  expect(props.reportError).toHaveBeenCalled();
});

test('check whether ruleDescriptions is set when getRuleDescription service resolved is successfully', async () => {
  expect(wrapper.state('ruleDescriptions')).toEqual([]);

  ruleService.getRuleDescription = jest.fn(() =>
    Promise.resolve(ruleDescriptionData[0])
  );

  await wrapper.instance().fetchDescription(data);

  expect(wrapper.state('ruleDescriptions')).toEqual([ruleDescriptionData[0]]);
});

test('check whether ruleDescriptions is reset and rules and currentRuleSetId is set when ruleset change is successfull', async () => {
  wrapper.setState({ ruleDescriptions: [ruleDescriptionData[0]] });

  ruleService.getRules = jest.fn(() => Promise.resolve({ data: rulesData }));

  await wrapper.instance().onChangeRuleSet({
    target: {
      value: 1,
    },
  });

  expect(wrapper.state('ruleDescriptions')).toEqual([]);
  expect(wrapper.state('rules')).toEqual(rulesData);
  expect(wrapper.state('currentRuleSetId')).toBe(1);
});

test("check whether rule's state change when activateRule service resolved successfully", async () => {
  wrapper.setState({ rules: rulesData });

  ruleService.activateRule = jest.fn(() => Promise.resolve());

  await wrapper.instance().handleStateChange(true, rulesData[0]);

  expect(wrapper.state('rules')[0].active).toBe(true);
  expect(wrapper.state('rules')[0].activeDefinedInCurrent).toBe(true);
});

test('check whether reportError is called when an error occured in activateRule service', async () => {
  ruleService.activateRule = jest.fn(() =>
    // eslint-disable-next-line prefer-promise-reject-errors
    Promise.reject({
      response: {},
    })
  );

  await wrapper.instance().handleStateChange(true, rulesData[1]);

  expect(props.reportError).toHaveBeenCalled();
});

test("check whether rule's state reverted when revertRule service resolved successfully", async () => {
  wrapper.setState({ rules: rulesData });

  ruleService.revertRule = jest.fn(() => Promise.resolve());
  ruleService.getRuleDescription = jest.fn(() =>
    Promise.resolve(ruleDescriptionData[1])
  );

  await wrapper.instance().handleStateReset(rulesData[1]);

  expect(wrapper.state('rules')[1].active).toBe(false);
  expect(wrapper.state('rules')[1].activeDefinedInCurrent).toBe(false);
});

test('check whether reportError is called when an error occured in revertRule service', async () => {
  ruleService.revertRule = jest.fn(() =>
    // eslint-disable-next-line prefer-promise-reject-errors
    Promise.reject({
      response: {},
    })
  );

  await wrapper.instance().handleStateReset(rulesData[1]);

  expect(props.reportError).toHaveBeenCalled();
});

test("check whether rule's parameter change when handleParamSet is called", async () => {
  wrapper.setState({
    rules: rulesData,
    ruleSets: ruleSets,
    currentRuleSetId: ruleSets[2].id,
    ruleDescriptions: [ruleDescriptionData[0]],
  });

  const data =
    ruleDescriptionData[0].userDescription.references[
      'SierraCiesApplyAllowDeadheadDuty0'
    ];
  const code = ruleDescriptionData[0].code;

  await wrapper.instance().handleParamSet('false', data, code);

  expect(
    wrapper.state('ruleDescriptions')[0].userDescription.references[
      'SierraCiesApplyAllowDeadheadDuty0'
    ].value
  ).toBe('false');
});

test("check whether rule's parameter reverted when revertParam service resolved successfully", async () => {
  wrapper.setState({
    rules: rulesData,
    ruleDescriptions: [ruleDescriptionData[0]],
  });

  ruleService.revertParam = jest.fn(() => Promise.resolve());
  ruleService.getRuleDescription = jest.fn(() =>
    Promise.resolve(ruleDescriptionData[0])
  );

  const data = ruleDescriptionData[0].userDescription[1];
  const code = ruleDescriptionData[0].code;
  const ruleset = ruleDescriptionData[0].ruleset;
  await wrapper.instance().handleParamReset(data, ruleset, code);

  expect(wrapper.state('rules')[0]).toEqual(rulesData[0]);
  expect(wrapper.state('ruleDescriptions')).toEqual([ruleDescriptionData[0]]);
});

test('check whether reportError is called when an error occured in revertParam service', async () => {
  ruleService.revertParam = jest.fn(() =>
    // eslint-disable-next-line prefer-promise-reject-errors
    Promise.reject({
      response: {},
    })
  );

  const data = ruleDescriptionData[0].userDescription[1];
  const code = ruleDescriptionData[0].code;
  const ruleset = ruleDescriptionData[0].ruleset;
  await wrapper.instance().handleParamReset(data, ruleset, code);

  expect(props.reportError).toHaveBeenCalled();
});

let readOnlyWrapper;
test('Rules component render correct in readOnly mode', () => {
  readOnlyWrapper = shallow(<Rules {...props} readOnly={true} />);
  expect(readOnlyWrapper).toMatchSnapshot();
});

test('Check whether state combo box is disabled in readOnly mode', () => {
  expect(readOnlyWrapper.instance().handleDisable('', {})).toBe(true);
});

test('Check whether state reset button is disabled in readOnly mode', () => {
  expect(readOnlyWrapper.instance().handleStateTooltipDisable('', {})).toBe(
    true
  );
});
