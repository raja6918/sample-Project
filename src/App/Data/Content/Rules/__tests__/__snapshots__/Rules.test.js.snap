// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rules component render correct 1`] = `
<Fragment>
  <AddHeader
    className=""
    disableAdd={false}
    editMode={false}
    name=""
    needAddButton={false}
    onClick={[Function]}
    openItemId="1"
    openItemName="Scenario test"
    redirectLink={
      <RedirectLink
        display="inline-block"
        margin="0 0 0 0px"
        to={
          Object {
            "pathname": "rules/rule-sets",
            "state": Object {
              "editMode": false,
              "openItemId": "1",
              "openItemName": "Scenario test",
              "readOnly": undefined,
            },
          }
        }
      >
        DATA.ruleSets.link
      </RedirectLink>
    }
    reportError={[MockFunction]}
    t={[Function]}
    userData={
      Object {
        "id": 1,
      }
    }
  />
  <ModalLoader
    background=""
    color="white"
    loaderRequired={true}
    onClose={[Function]}
    open={false}
    subtitle=""
    title="DATA.rules.loader"
  />
  <Container>
    <WithStyles(ForwardRef(Grid))
      alignItems="center"
      className="container-cls"
      container={true}
      spacing={1}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        DATA.rules.inputSelector
         :
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <WithStyles(ForwardRef(Select))
          className="ruleset-select"
          inputProps={
            Object {
              "id": "",
              "name": "",
              "required": false,
            }
          }
          onChange={[Function]}
          value={null}
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <DataNotFound
      style={Object {}}
      text="GLOBAL.dataLoading.message"
    />
  </Container>
</Fragment>
`;

exports[`Rules component render correct in readOnly mode 1`] = `
<Fragment>
  <AddHeader
    className=""
    disableAdd={false}
    editMode={false}
    name=""
    needAddButton={false}
    onClick={[Function]}
    openItemId="1"
    openItemName="Scenario test"
    readOnly={true}
    redirectLink={
      <RedirectLink
        display="inline-block"
        margin="0 0 0 0px"
        to={
          Object {
            "pathname": "rules/rule-sets",
            "state": Object {
              "editMode": false,
              "openItemId": "1",
              "openItemName": "Scenario test",
              "readOnly": true,
            },
          }
        }
      >
        DATA.ruleSets.link
      </RedirectLink>
    }
    reportError={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "error": Object {
                "response": Object {},
              },
            },
          ],
          Array [
            Object {
              "error": Object {
                "response": Object {},
              },
            },
          ],
          Array [
            Object {
              "error": Object {
                "response": Object {},
              },
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    t={[Function]}
    userData={
      Object {
        "id": 1,
      }
    }
  />
  <ModalLoader
    background=""
    color="white"
    loaderRequired={true}
    onClose={[Function]}
    open={false}
    subtitle=""
    title="DATA.rules.loader"
  />
  <Container>
    <WithStyles(ForwardRef(Grid))
      alignItems="center"
      className="container-cls"
      container={true}
      spacing={1}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        DATA.rules.inputSelector
         :
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <WithStyles(ForwardRef(Select))
          className="ruleset-select"
          inputProps={
            Object {
              "id": "",
              "name": "",
              "required": false,
            }
          }
          onChange={[Function]}
          value={1}
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <DataNotFound
      style={Object {}}
      text="GLOBAL.dataLoading.message"
    />
  </Container>
</Fragment>
`;
